src/io/linux.zig

```diff
@@ -112,14 +112,11 @@ pub const IO = struct {
         }
 
         // Run completions only after all completions have been flushed:
-        // Loop on a copy of the linked list, having reset the list first, so that any synchronous
-        // append on running a completion is executed only the next time round the event loop,
-        // without creating an infinite loop.
-        {
-            var copy = self.completed;
-            self.completed = .{};
-            while (copy.pop()) |completion| completion.complete();
-        }
+        // Loop until all completions are processed. Calls to complete() may queue more work
+        // and extend the duration of the loop, but this is fine as it 1) executes completions
+        // that become ready without going through another syscall from flush_submissions() and
+        // 2) potentially queues more SQEs to take advantage more of the next flush_submissions().
+        while (self.completed.pop()) |completion| completion.complete();
```

Comment:
This change makes sense! But, unrelated to the PR, I am wondering about the submit/complete ordering here. The comment says that we submit first to avoid starvation, but can't we just always submit in fifo to avoid that? 

```diff
diff --git a/src/io/linux.zig b/src/io/linux.zig
index 3fd9bf50..3fd4f9b5 100644
--- a/src/io/linux.zig
+++ b/src/io/linux.zig
@@ -101,6 +101,13 @@ pub const IO = struct {
         // We can now just peek for any CQEs without waiting and without another syscall:
         try self.flush_completions(0, timeouts, etime);
 
+        // Run completions only after all completions have been flushed:
+        // Loop until all completions are processed. Calls to complete() may queue more work
+        // and extend the duration of the loop, but this is fine as it 1) executes completions
+        // that become ready without going through another syscall from flush_submissions() and
+        // 2) potentially queues more SQEs to take advantage more of the next flush_submissions().
+        while (self.completed.pop()) |completion| completion.complete();
+
         // The SQE array is empty from flush_submissions(). Fill it up with unqueued completions.
         // This runs before `self.completed` is flushed below to prevent new IO from reserving SQE
         // slots and potentially starving those in `self.unqueued`.
@@ -111,13 +118,6 @@ pub const IO = struct {
             while (copy.pop()) |completion| self.enqueue(completion);
         }
 
-        // Run completions only after all completions have been flushed:
-        // Loop until all completions are processed. Calls to complete() may queue more work
-        // and extend the duration of the loop, but this is fine as it 1) executes completions
-        // that become ready without going through another syscall from flush_submissions() and
-        // 2) potentially queues more SQEs to take advantage more of the next flush_submissions().
-        while (self.completed.pop()) |completion| completion.complete();
-
         // At this point, unqueued could have completions either by 1) those who didn't get an SQE
         // during the popping of unqueued or 2) completion.complete() which start new IO. These
         // unqueued completions will get priority to acquiring SQEs on the next flush().
@@ -175,13 +175,11 @@ pub const IO = struct {
     }
 
     fn enqueue(self: *IO, completion: *Completion) void {
+        self.unqueued.push(completion);
         const sqe = self.ring.get_sqe() catch |err| switch (err) {
-            error.SubmissionQueueFull => {
-                self.unqueued.push(completion);
-                return;
-            },
+            error.SubmissionQueueFull => return,
         };
-        completion.prep(sqe);
+        self.unqueued.pop().?.prep(sqe);
     }
 
     /// This struct holds the data needed for a single io_uring operation
```

Reply:
TBH, that looks like a much nicer solution over #464 and having to worry about when to do unqueued re-submission.

Reply:
Actually, realized it wouldn't handle re-assigning SQEs to *pending* unqueued when they become free in flush() without a similar loop there.